### PR TITLE
DOC: Add missing docstring error to validate_docstrings.py

### LINKED ISSUE #23648: DOC: Add validation error when a docstring is missing
We have the script `scripts/validate_docstrings.py` that checks several formatting errors in the docstrings.

Currently, when a function, method... has no docstring at all, it fails few of the checks, but we don't have one specific to it. See for example:
```
$ ./scripts/validate_docstrings.py --prefix=pandas.Series.imag
pandas.Series.imag: Docstring text (summary) should start in the line immediately after the opening quotes (not in the same line, or leaving a blank line in between)
pandas.Series.imag: Closing quotes should be placed in the line after the last text in the docstring (do not close the quotes in the same line as the text, or leave a blank line between the last text and the quotes)
pandas.Series.imag: No summary found (a short summary in a single line should be present at the beginning of the docstring)
```

What would be useful is to add a new check, that reports that the function does not have a docstring. And in the previous errors, check when the docstring is empty, and do not report them. So the previous call should report a single error with a text like `pandas.Series.imag: No docstring found`.

### PR DESCRIPTION
**Pull request by avolkov**
This is an update of a previous pull request -- https://github.com/pandas-dev/pandas/pull/23669
with all unrelated commits and merges removed and and updates added from code review of that pull request.

- [x] closes #23648
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by datapythonista** [comment]
This addresses the same issue as #23668. Can you discuss with @iway1 to see who wants to finish the work? Both PRs require some work, I don't like having all the code indented, or having the big `return` twice. I think the best would be having a separate function for the checks that just returns the errors and warnings, so we can have at the beginning:
```
if doc.missing_docstring:
    errs.append(error("GL08"))
    return errs, wrns
```

**Comment by avolkov** [comment]
@iway1 I can finish working on the PR. I think I got the idea how to refactor the code into a separate function that handles only errs and wrns while ``validate_one`` would only handle conversion of those into output dictionary.
In this version the logic still has return statement at the top, however it is limited in scope only to the logic related to errs and wrns.

**Comment by datapythonista** [comment]
@avolkov do you have time to update based on feedback and fix the conflicts?

**Comment by avolkov** [comment]
yes, I'll do it today

File: scripts/tests/test_validate_docstrings.py [hunk 1]
```diff
@@ -720,6 +726,13 @@ def test_bad_class(self):
         assert isinstance(errors, list)
         assert errors
 
+    @capture_stderr
+    def test_no_docstring(self):
+        errors = validate_one(self._import_path(
+            klass='NoDocstrings'))['errors']
+        assert isinstance(errors, list)
+        assert errors
```
[the remark this answers is not in the document]
Comment (avolkov) [reply scripts/tests/test_validate_docstrings.py]: ok, I move the method to to BadGenericDoc string and remove this test.

File: scripts/tests/test_validate_docstrings.py [hunk 2]
```diff
@@ -812,6 +817,7 @@ def test_bad_generic_functions(self, func):
           'E226 missing whitespace around arithmetic operator',)),
         ('BadExamples', 'missing_whitespace_after_comma',
          ("flake8 error: E231 missing whitespace after ',' (3 times)",)),
+
```
Comment (datapythonista) [inline scripts/tests/test_validate_docstrings.py]: try in general to avoid having changes unrelated to the PR

File: scripts/validate_docstrings.py [hunk 3]
```diff
@@ -550,11 +551,16 @@ def validate_one(func_name):
     func_name : function
         Function whose docstring will be evaluated (e.g. pandas.read_csv).
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: This blank line is not needed.

File: scripts/validate_docstrings.py [hunk 4]
```diff
@@ -550,11 +551,16 @@ def validate_one(func_name):
...
 
+    doc : Docstring
+        A Docstring object with the given function name
+
+
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: Leave just one blank between sections. The CI will fail soon because of this

File: scripts/validate_docstrings.py [hunk 5]
```diff
@@ -550,11 +551,16 @@ def validate_one(func_name):
...
     Returns
     -------
-    dict
-        A dictionary containing all the information obtained from validating
-        the docstring.
+
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: blank line not needed

File: scripts/validate_docstrings.py [hunk 6]
```diff
@@ -550,11 +551,16 @@ def validate_one(func_name):
...
+    tuple
+        A tuple of lists containing all the errors, warnings and example_errors
+        that occurred during validation
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: For tuples, the format would be instead:
```
errors : list of tuple
    Errors description.
warnings : list of tuple
    ...
```

File: scripts/validate_docstrings.py [hunk 7]
```diff
@@ -627,8 +636,8 @@ def validate_one(func_name):
             errs.append(error('SS03'))
         if doc.summary != doc.summary.lstrip():
             errs.append(error('SS04'))
-        elif (doc.is_function_or_method
-                and doc.summary.split(' ')[0][-1] == 's'):
+        elif (doc.is_function_or_method and
+                doc.summary.split(' ')[0][-1] == 's'):
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: As said, try to avoid unrelated changes, they make things more difficult. And in this case, if I remember correctly, the Python standard is to have the binary operator after the line break, as it was before.

File: scripts/validate_docstrings.py [hunk 8]
```diff
@@ -704,7 +713,26 @@ def validate_one(func_name):
         for wrong_import in ('numpy', 'pandas'):
             if 'import {}'.format(wrong_import) in examples_source_code:
                 errs.append(error('EX04', imported_library=wrong_import))
+    return errs, wrns, examples_errs
+
+
+def validate_one(func_name):
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: I think the names of the functions could be better. Not sure what would be the clearest, but if we leave this as `validate_one`, it would make things simpler to call the other `perform_validation`, `validation_conditions` or something like that. Or may be that one should be `validate_one`, and this should be `get_validation_data`, or something similar.

Not sure what's best, but the current names don't tell much about what each one does.

File: scripts/validate_docstrings.py [hunk 9]
```diff
@@ -541,7 +542,7 @@ def validate_pep8(self):
         yield from application.guide.stats.statistics_for('')
 
 
-def validate_one(func_name):
+def validate_errs_wrns(func_name, doc):
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: `doc` should have an attribute with the `func_name`, I think it's redundant here.

**Review by datapythonista — COMMENTED** [review]
Looks good, I'd just make few minor formatting and readability changes. And the PR also requires a rebase.

File: scripts/validate_docstrings.py (hunk 7 above)
Comment (avolkov) [reply scripts/validate_docstrings.py]: Sorry I was looking at the related code and my editor highlighted PEP8 warning (W503): line break before binary operator. I'll revert this to follow accepted convention.

File: scripts/validate_docstrings.py (hunk 8 above)
Comment (avolkov) [reply scripts/validate_docstrings.py]: I like `get_validation_data` it seems to be the most descriptive.

**Review by datapythonista — APPROVED** [review]
lgtm, thanks for the work and the updates @avolkov

**Comment by datapythonista** [comment]
Can you please rebase? there is a conflict with master

**Comment by datapythonista** [comment]
@avolkov something went wrong when merging with master, and you've got plenty of unrelated changes. Can you fix it please?

**Comment by avolkov** [comment]
@datapythonista It looks like I didn't properly do git rebase and the tests I wrote are failing. I'm working on it.

**Comment by datapythonista** [comment]
When we ask for a rebase, it's not literally a rebase but a `git fetch upstream && git merge upstream/master`.

If you don't find a better way, `git fetch upstream && git reset --soft upstream/master` usually works for me (it's explained in more detail here: https://datapythonista.github.io/blog/useful-git-commands.html)

**Comment by avolkov** [comment]
There was an issue with the test ( I didn't properly remove test code for NoDoctsring class that since have been removed).  All the checks seem to be passing now.

**Comment by datapythonista** [comment]
I still see 105 commits and a huge diff. Couldn't you fix that?

**Comment by avolkov** [comment]
Oh, ok, I see, there's another issue where I need to follow your rebase procedure. I'll do that.

**Comment by datapythonista** [comment]
rebased

**Comment by datapythonista** [comment]
@jreback this should be ready

**Comment by jreback** [comment]
thanks @avolkov
